## Road Accidents Analysis
****
#### Objective:
The goal of this project is to analyze key factors contributing to road accidents and road safety issues. The project involves cleaning and sampling the dataset using Python (Pandas), followed by building an interactive Excel dashboard to present insights visually.

#### Data Source:
- [Road Accident Dataset (Kaggle)](https://www.kaggle.com/datasets/abdulmannann/road-accidents-csv)
- Download data using above link and save in same folder as this notebook.

#### Dataset Description
- This dataset contains detailed records of road accidents from a specific region over a given period. It includes attributes such as location, time, weather conditions, road surface, light conditions, vehicle types, and casualties.

---

### Step 1: Data Understanding
- Load the dataset using Pandas and perform initial exploration using methods like .info(), .describe(), .value_counts().
- Identify redundant columns, missing values, duplicates, and inconsistent entries.

In [5]:
# Importing pandas
import pandas as pd

# Loading Dataset into dataframe
df=pd.read_excel('Road_Accident_Data.xlsx')

# Displaying 1st 5 rows 
df.head()

Accident_Index Accident Date Day_of_Week          Junction_Control  \
0  200901BS70001    2021-01-01    Thursday  Give way or uncontrolled   
1  200901BS70002    2021-01-05      Monday  Give way or uncontrolled   
2  200901BS70003    2021-01-04      Sunday  Give way or uncontrolled   
3  200901BS70004    2021-01-05      Monday       Auto traffic signal   
4  200901BS70005    2021-01-06     Tuesday       Auto traffic signal   

           Junction_Detail Accident_Severity   Latitude  \
0  T or staggered junction           Serious  51.512273   
1               Crossroads           Serious  51.514399   
2  T or staggered junction            Slight  51.486668   
3  T or staggered junction           Serious  51.507804   
4               Crossroads           Serious  51.482076   

        Light_Conditions Local_Authority_(District) Carriageway_Hazards  ...  \
0               Daylight     Kensington and Chelsea                 NaN  ...   
1               Daylight     Kensington and Chelsea                 NaN  ...   
2               Daylight     Kensington and Chelsea                 NaN  ...   
3               Daylight     Kensington and Chelsea                 NaN  ...   
4  Darkness - lights lit     Kensington and Chelsea                 NaN  ...   

   Number_of_Casualties  Number_of_Vehicles         Police_Force  \
0                     1                   2  Metropolitan Police   
1                    11                   2  Metropolitan Police   
2                     1                   2  Metropolitan Police   
3                     1                   2  Metropolitan Police   
4                     1                   2  Metropolitan Police   

  Road_Surface_Conditions           Road_Type Speed_limit      Time  \
0                     Dry      One way street          30  15:11:00   
1             Wet or damp  Single carriageway          30  10:59:00   
2                     Dry  Single carriageway          30  14:19:00   
3            Frost or ice  Single carriageway          30  08:10:00   
4                     Dry  Single carriageway          30  17:25:00   

  Urban_or_Rural_Area  Weather_Conditions           Vehicle_Type  
0               Urban  Fine no high winds                    Car  
1               Urban  Fine no high winds  Taxi/Private hire car  
2               Urban  Fine no high winds  Taxi/Private hire car  
3               Urban               Other  Motorcycle over 500cc  
4               Urban  Fine no high winds                    Car  

[5 rows x 21 columns]

In [6]:
# Checking shape
df.shape

(307973, 21)

In [7]:
# Basic info of dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307973 entries, 0 to 307972
Data columns (total 21 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   Accident_Index              307973 non-null  object        
 1   Accident Date               307973 non-null  datetime64[ns]
 2   Day_of_Week                 307973 non-null  object        
 3   Junction_Control            307973 non-null  object        
 4   Junction_Detail             307973 non-null  object        
 5   Accident_Severity           307973 non-null  object        
 6   Latitude                    307973 non-null  float64       
 7   Light_Conditions            307973 non-null  object        
 8   Local_Authority_(District)  307973 non-null  object        
 9   Carriageway_Hazards         5424 non-null    object        
 10  Longitude                   307973 non-null  float64       
 11  Number_of_Casualties        307973 non-

In [8]:
# To get descriptive statistics of numerical or datetime columns
df.describe()

Accident Date       Latitude      Longitude  \
count                         307973  307973.000000  307973.000000   
mean   2021-12-23 22:19:39.804722944      52.487005      -1.368884   
min              2021-01-01 00:00:00      49.914488      -7.516225   
25%              2021-06-28 00:00:00      51.485248      -2.247937   
50%              2021-12-08 00:00:00      52.225943      -1.349258   
75%              2022-06-25 00:00:00      53.415517      -0.206810   
max              2022-12-31 00:00:00      60.598055       1.759398   
std                              NaN       1.339011       1.356092   

       Number_of_Casualties  Number_of_Vehicles    Speed_limit  
count         307973.000000       307973.000000  307973.000000  
mean               1.356882            1.829063      38.866037  
min                1.000000            1.000000      10.000000  
25%                1.000000            1.000000      30.000000  
50%                1.000000            2.000000      30.000000  
75%                1.000000            2.000000      50.000000  
max               48.000000           32.000000      70.000000  
std                0.815857            0.710477      14.032933

In [9]:
# To get descriptive statistics of other columns
df.describe(exclude=['int','float','datetime64'])

Accident_Index Day_of_Week          Junction_Control  \
count           307973      307973                    307973   
unique          197644           7                         7   
top      2010000000000      Friday  Give way or uncontrolled   
freq            110304       50529                    150045   

                            Junction_Detail Accident_Severity  \
count                                307973            307973   
unique                                    9                 4   
top     Not at junction or within 20 metres            Slight   
freq                                 123094            263280   

       Light_Conditions Local_Authority_(District)   Carriageway_Hazards  \
count            307973                     307973                  5424   
unique                5                        422                     5   
top            Daylight                 Birmingham  Other object on road   
freq             227286                       6165                  2243   

               Police_Force Road_Surface_Conditions           Road_Type  \
count                307973                  307656              306439   
unique                   51                       5                   5   
top     Metropolitan Police                     Dry  Single carriageway   
freq                  46789                  208967              230612   

            Time Urban_or_Rural_Area  Weather_Conditions Vehicle_Type  
count     307956              307973              301916       307973  
unique      1439                   2                   8           15  
top     17:00:00               Urban  Fine no high winds          Car  
freq        2933              198532              244496       239794

In [10]:
# Duplicated rows
df[df.duplicated(keep=False)]

Accident_Index Accident Date Day_of_Week              Junction_Control  \
50482  2010000000000    2021-01-06     Tuesday  Data missing or out of range   
50483  2010000000000    2021-01-06     Tuesday  Data missing or out of range   

                           Junction_Detail Accident_Severity   Latitude  \
50482  Not at junction or within 20 metres            Slight  53.648498   
50483  Not at junction or within 20 metres            Slight  53.648498   

      Light_Conditions Local_Authority_(District) Carriageway_Hazards  ...  \
50482         Daylight                 Calderdale                 NaN  ...   
50483         Daylight                 Calderdale                 NaN  ...   

       Number_of_Casualties  Number_of_Vehicles    Police_Force  \
50482                     1                   1  West Yorkshire   
50483                     1                   1  West Yorkshire   

      Road_Surface_Conditions           Road_Type Speed_limit      Time  \
50482            Frost or ice  Single carriageway          50  16:35:00   
50483            Frost or ice  Single carriageway          50  16:35:00   

      Urban_or_Rural_Area  Weather_Conditions Vehicle_Type  
50482               Rural  Fine no high winds          Car  
50483               Rural  Fine no high winds          Car  

[2 rows x 21 columns]

In [11]:
# Total Null values in each column
df.isnull().sum()

Accident_Index                     0
Accident Date                      0
Day_of_Week                        0
Junction_Control                   0
Junction_Detail                    0
Accident_Severity                  0
Latitude                           0
Light_Conditions                   0
Local_Authority_(District)         0
Carriageway_Hazards           302549
Longitude                          0
Number_of_Casualties               0
Number_of_Vehicles                 0
Police_Force                       0
Road_Surface_Conditions          317
Road_Type                       1534
Speed_limit                        0
Time                              17
Urban_or_Rural_Area                0
Weather_Conditions              6057
Vehicle_Type                       0
dtype: int64

In [12]:
# Percentage of Null values in each column
(df.isnull().sum()*100/len(df)).round(2)

Accident_Index                 0.00
Accident Date                  0.00
Day_of_Week                    0.00
Junction_Control               0.00
Junction_Detail                0.00
Accident_Severity              0.00
Latitude                       0.00
Light_Conditions               0.00
Local_Authority_(District)     0.00
Carriageway_Hazards           98.24
Longitude                      0.00
Number_of_Casualties           0.00
Number_of_Vehicles             0.00
Police_Force                   0.00
Road_Surface_Conditions        0.10
Road_Type                      0.50
Speed_limit                    0.00
Time                           0.01
Urban_or_Rural_Area            0.00
Weather_Conditions             1.97
Vehicle_Type                   0.00
dtype: float64

In [13]:
# To find typos in categories 
for col_name in df.columns:
    print(col_name,':\n',df[col_name].unique(),'\n')

Accident_Index :
 ['200901BS70001' '200901BS70002' '200901BS70003' ... '201091NM01935'
 '201091NM01964' '201091NM02142'] 

Accident Date :
 <DatetimeArray>
['2021-01-01 00:00:00', '2021-01-05 00:00:00', '2021-01-04 00:00:00',
 '2021-01-06 00:00:00', '2021-01-08 00:00:00', '2021-01-02 00:00:00',
 '2021-01-07 00:00:00', '2021-01-10 00:00:00', '2021-01-16 00:00:00',
 '2021-01-12 00:00:00',
 ...
 '2022-09-13 00:00:00', '2022-10-31 00:00:00', '2022-11-21 00:00:00',
 '2022-12-02 00:00:00', '2022-12-18 00:00:00', '2022-12-29 00:00:00',
 '2022-12-30 00:00:00', '2022-12-26 00:00:00', '2022-07-25 00:00:00',
 '2022-12-25 00:00:00']
Length: 730, dtype: datetime64[ns] 

Day_of_Week :
 ['Thursday' 'Monday' 'Sunday' 'Tuesday' 'Friday' 'Wednesday' 'Saturday'] 

Junction_Control :
 ['Give way or uncontrolled' 'Auto traffic signal'
 'Data missing or out of range' 'Authorised person' 'Stop sign'
 'Not at junction or within 20 metres' 'Auto traffic sigl'] 

Junction_Detail :
 ['T or staggered junction' 'C

In [14]:
# Missing values in Junction_Control
(df['Junction_Control']=='Data missing or out of range').sum()

98056

In [15]:
# Percentage of missing values in Junction_Control
(df['Junction_Control']=='Data missing or out of range').sum() * 100 /len(df)

31.839154731096556

#### Observations (Step 1: Data Understanding)

- The dataset has 21 columns and 307973 records
  
- **Redundant column**:
   - Carriageway_Hazards is not useful for analysis as it has almost no variance (>98% null values or None type) -> Drop 

- **Duplicate rows**:
   - 1 duplicate row is there -> Drop
     
- **Missing Values**:
   - Road_Surface_Conditions -> 0.1%
   - Road_Type -> 0.5%
   - Time -> 0.01%
   - Weather_Conditions -> 1.97%
   - Junction_Control has 31% unknown but not null -> Keep as is for now.
     
- **Data Entry Issues or Typos**:
   - Accident_Index -> Repeated values indicate possible entry errors, but it will only be used for counting, so can be retained.
   - Junction_Control: 'Auto traffic sigl' -> Replace it with 'Auto traffic signal'
   - Accident_Severity: 'Fetal' -> Replace it with 'Fatal'

---
### Step 2: Data Cleaning
- Drop redundant column
- Handle duplicates & missing values
- Handle inconsistent entries.

In [18]:
# Making copy to work safely
df_copy = df.copy()

# Drop Carriageway_Hazards
df_copy.drop(columns='Carriageway_Hazards',inplace=True)

# Drop duplicates
df_copy.drop_duplicates(inplace=True)

# Drop rows having null values
df_copy.dropna(inplace=True)

# Handling inconsistent entries
df_copy.Junction_Control=df_copy.Junction_Control.replace('Auto traffic sigl','Auto traffic signal')
df_copy.Accident_Severity=df_copy.Accident_Severity.replace('Fetal','Fatal')

# Confirm changes
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 300494 entries, 0 to 307972
Data columns (total 20 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   Accident_Index              300494 non-null  object        
 1   Accident Date               300494 non-null  datetime64[ns]
 2   Day_of_Week                 300494 non-null  object        
 3   Junction_Control            300494 non-null  object        
 4   Junction_Detail             300494 non-null  object        
 5   Accident_Severity           300494 non-null  object        
 6   Latitude                    300494 non-null  float64       
 7   Light_Conditions            300494 non-null  object        
 8   Local_Authority_(District)  300494 non-null  object        
 9   Longitude                   300494 non-null  float64       
 10  Number_of_Casualties        300494 non-null  int64         
 11  Number_of_Vehicles          300494 non-null 

In [19]:
# % records dropped
(307973-300494)*100/307973

2.428459637695512

In [20]:
# Check duplicates
df_copy[df_copy.duplicated()]

Empty DataFrame
Columns: [Accident_Index, Accident Date, Day_of_Week, Junction_Control, Junction_Detail, Accident_Severity, Latitude, Light_Conditions, Local_Authority_(District), Longitude, Number_of_Casualties, Number_of_Vehicles, Police_Force, Road_Surface_Conditions, Road_Type, Speed_limit, Time, Urban_or_Rural_Area, Weather_Conditions, Vehicle_Type]
Index: []

In [21]:
# Check null
df_copy.isnull().sum()

Accident_Index                0
Accident Date                 0
Day_of_Week                   0
Junction_Control              0
Junction_Detail               0
Accident_Severity             0
Latitude                      0
Light_Conditions              0
Local_Authority_(District)    0
Longitude                     0
Number_of_Casualties          0
Number_of_Vehicles            0
Police_Force                  0
Road_Surface_Conditions       0
Road_Type                     0
Speed_limit                   0
Time                          0
Urban_or_Rural_Area           0
Weather_Conditions            0
Vehicle_Type                  0
dtype: int64

In [22]:
# Check Junction_Control for typos
df_copy.Junction_Control.unique()

array(['Give way or uncontrolled', 'Auto traffic signal',
       'Data missing or out of range', 'Authorised person', 'Stop sign',
       'Not at junction or within 20 metres'], dtype=object)

In [23]:
# Check Accident_Severity for typos
df_copy.Accident_Severity.unique()

array(['Serious', 'Slight', 'Fatal'], dtype=object)

In [24]:
# Assigning copy back to original dataframe
df=df_copy

---
### Step 3: Data Sampling and Saving to excel
- Take 5k rows as sample from cleaned data and save to excel for dashboarding

In [26]:
# Sample 5,000 rows randomly
sample_df = df.sample(n=5000, random_state=18)

# Save the sampled data to a new Excel file
sample_df.to_excel("Road_Accident_Cleaned_Data.xlsx", index=False)

---
#### Excel Dashboard Overview (Road_Accident_Analysis.xlsx):
- 4 KPIs: Total Accidents, Casualties, Vehicles, Avg. Casualties per Accident
- 6 Visuals: Month-wise trend, Hour-wise trend, Accidents By Severity, Accidents By Vehicle type, Accidents By Road surface, Accidents By Junction control, 
- 3 Slicers: Year, Region, Severity